README.md
# get_photo

[![Pub](https://img.shields.io/pub/v/get_photo.svg)](https://pub.dev/packages/get_photo)

A new photo selection plug-in UI, packaged with the photo_manager interface, supports IOS, Android, macOS; only supports selection of pictures.

一款新的照片选择插件UI，使用photo_manager 接口封装，支持IOS, Android, macOS；仅支持选择图片。

## Screenshots

![lineopts example screenshot](screenshots/11.png)

## Example

```dart
Uint8List? _image;

GetPhoto(
  crossAxisCount: 4,
  lang: {
    'Recent': '所有图片',
    'Camera': '相机',
    'WeiXin': '微信',
  },
  onTap: (v) async {
    _image = await v.originBytes;
    //print((await v.titleAsync));
    //setState(() {});
  },
),
```

## Customization

Here is a list of properties available to customize your widget:

| Name | Type | Description |
|-----|-----|------|
|crossAxisCount| int | Display quantity in one line |
|onTap| ValueChanged<AssetEntity> | Choose to call back a parameter of type AssetEntity|
|lang| Map | Replace the default album name |
|showAlbumsList| bool | Show album list or all pictures by default |

| 名称 | 类型 | 描述 |
|-----|-----|------|
|crossAxisCount| int | 一行显示数量 |
|onTap| ValueChanged<AssetEntity> | 选择回调一个AssetEntity类型的参数|
|lang| Map | 替换默认相册名称 |
|showAlbumsList| bool | 默认显示相册列表还是全部图片 |

## FAQ

使用glide
在/android/build.gradle 增加

```
rootProject.allprojects {

    subprojects {
        project.configurations.all {
            resolutionStrategy.eachDependency { details ->
                if (details.requested.group == 'com.github.bumptech.glide'
                        && details.requested.name.contains('glide')) {
                    details.useVersion '4.11.0'
                }
            }
        }
    }
}
```

在/android/app/build.gradle 增加

```
+ apply plugin: 'kotlin-kapt'

dependencies {
    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
   + implementation 'com.github.bumptech.glide:glide:4.11.0'
   + kapt 'com.github.bumptech.glide:compiler:4.11.0'
}

```

在/android/app/src/main/kotlin/com/xx/xx/

新建 AppGlideModule.kt 拷贝以下内容保存

```
package com.example.getphoto

import com.bumptech.glide.annotation.GlideModule
import com.bumptech.glide.module.AppGlideModule
@GlideModule
class MyAppGlideModule : AppGlideModule(){
}
```


CHANGELOG.md
## [1.0.0] 

* publish


example/lib/main.dart
import 'dart:typed_data';

import 'package:flutter/material.dart';
import 'package:get_photo/get_photo.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  Uint8List? _image;

  void _getImage() async {
    Navigator.of(context).push(
      MaterialPageRoute(
        builder: (context) => Scaffold(
          body: GetPhoto(
            crossAxisCount: 4,
            // showType: ShowType.all,
            lang: {
              'Recent': '所有图片',
              'Camera': '相机',
              'WeiXin': '微信',
            },
            onTap: (v) async {
              _image = await v.originBytes;
              print((await v.titleAsync));
              setState(() {});
            },
          ),
        ),
        fullscreenDialog: true,
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        centerTitle: true,
        title: Text('getPhoto demo'),
      ),
      body: Center(
        child: _image == null
            ? MaterialButton(
                onPressed: () => _getImage(),
                child: Icon(Icons.image),
                color: Colors.blue,
              )
            : Image.memory(
                _image!,
                fit: BoxFit.cover,
                width: 200.0,
              ),
      ),

      floatingActionButton: FloatingActionButton(
        onPressed: _getImage,
        tooltip: 'getImage',
        child: Icon(Icons.image),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}


lib/src/albumsList.dart
import 'dart:typed_data';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:photo_manager/photo_manager.dart';

class AlbumsList extends StatefulWidget {
  ///albums list
  late final List<AssetPathEntity> albumsList;

  ///albums list onTap
  final ValueChanged<AssetPathEntity> onTap;

  AlbumsList(
    this.albumsList, {
    required this.onTap,
  });
  @override
  _AlbumsListState createState() => _AlbumsListState();
}

class _AlbumsListState extends State<AlbumsList> {
  List<AssetPathEntity> albums = [];
  List<Uint8List?> albumsThumb = [];
  @override
  void initState() {
    super.initState();
    _albumsList();
  }

  @override
  void dispose() {
    super.dispose();
  }

  Future<void> _albumsList() async {
    List<AssetPathEntity> _albums = widget.albumsList;
    //sort by assetCount
    _albums.sort((a, b) => b.assetCount.compareTo(a.assetCount));
    albums.addAll(_albums);

    for (var item in albums) {
      final assetList = await item.getAssetListRange(start: 0, end: 1);
      if (assetList.isNotEmpty) {
        final Uint8List? _albums = await thumbDataWithSize(assetList.first);
        albumsThumb.add(_albums);
        if (mounted) setState(() {});
      }
    }
  }

  Future<Uint8List?> thumbDataWithSize(AssetEntity asset) async {
    return await asset.thumbData;
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      color: Colors.black,
      child: ListView.separated(
        itemBuilder: (BuildContext context, int index) {
          return imageList(index);
        },
        separatorBuilder: (BuildContext context, int index) {
          return Divider(
            height: 0,
            color: Colors.grey[800],
          );
        },
        itemCount: widget.albumsList.length,
        physics: BouncingScrollPhysics(),
      ),
    );
  }

  Widget emptyList() {
    return ListTile(
      contentPadding:
          const EdgeInsets.symmetric(vertical: 5.0, horizontal: 5.0),
      leading: Container(
        width: 50,
        height: 50,
        color: Colors.grey,
        margin: EdgeInsets.all(1.0),
      ),
      title: Container(
        constraints: BoxConstraints.tightForFinite(width: 100),
        child: Container(
          height: 30,
          color: Color(0xff4C4A48),
          margin: EdgeInsets.only(right: 150.0),
        ),
      ),
    );
  }

  Widget _leading(int index) {
    if (albumsThumb.length >= index + 1) {
      return Container(
        width: 50,
        height: 50,
        color: Colors.grey,
        margin: EdgeInsets.all(1.0),
        child: Image.memory(
          albumsThumb[index]!,
          fit: BoxFit.cover,
        ),
      );
    }

    // albumsThumb
    return Container(
      width: 50,
      height: 50,
      color: Colors.grey,
      margin: EdgeInsets.all(1.0),
    );
  }

  Widget imageList(int index) {
    return ListTile(
      contentPadding:
          const EdgeInsets.symmetric(vertical: 5.0, horizontal: 5.0),
      leading: _leading(index),
      title: RichText(
        text: TextSpan(children: <InlineSpan>[
          TextSpan(
            text: '${albums[index].name}',
            style: TextStyle(
              color: Colors.white,
              fontSize: 18.0,
            ),
          ),
          TextSpan(
            text: ' ( ${albums[index].assetCount} )',
            style: TextStyle(
              color: Colors.grey[600],
            ),
          ),
        ]),
      ),
      onTap: () {
        widget.onTap(albums[index]);
      },
    );
  }
}


lib/src/getPhoto.dart
import 'dart:typed_data';

import 'package:flutter/material.dart';
import 'package:photo_manager/photo_manager.dart';

import 'albumsList.dart';

/// wechat photo selection can only choose pictures
///
/// 仿微信照片选择，只能选择图片
class GetPhoto extends StatefulWidget {
  GetPhoto({
    this.onTap,
    this.crossAxisCount = 4,
    this.lang = const {},
    this.showAlbumsList = false,
  });

  ///Select the callback return asse tentity type
  ///
  ///选择回调，返回AssetEntity类型
  final ValueChanged<AssetEntity>? onTap;

  ///list one line display quantity
  ///
  ///列表一行显示数量
  final int crossAxisCount;

  ///localization 本地化
  ///```dart
  ///lang:{
  ///   'Recent': '所有图片',
  ///   'Camera': '相机',
  ///   'WeiXin': '微信',
  ///}
  ///```
  final Map lang;

  ///whether the album list is displayed by default the default no
  ///
  ///默认是否显示相册列表，默认：否
  final bool showAlbumsList;
  @override
  _GetPhotoState createState() => _GetPhotoState();
}

class _GetPhotoState extends State<GetPhoto> {
  List<Widget> _mediaList = [];
  List<AssetEntity> media = [];
  List<AssetPathEntity>? albums;
  int currentPage = 0;
  int lastPage = -1;
  bool _showAlbumsList = false;
  AssetPathEntity? currentAlbums;
  bool isDark = false;
  @override
  void initState() {
    super.initState();
    _showAlbumsList = widget.showAlbumsList;
    _init();
  }

  _loadMore(ScrollNotification scroll) {
    if (scroll.metrics.pixels / scroll.metrics.maxScrollExtent > 0.33) {
      if (currentPage != lastPage) {
        lastPage = currentPage;
        imageList();
      }
    }
  }

  _init() async {
    var result = await PhotoManager.requestPermission();

    if (result) {
      albums = await PhotoManager.getAssetPathList(
        onlyAll: false,
        type: RequestType.image,
      );
      if (albums != null) {
        for (var item in albums!) {
          item.name = _albumsName(item.name);
        }
        currentAlbums = albums![0];
        imageList();
      }
    } else {
      PhotoManager.openSetting();
    }
  }

  String _albumsName(String name) {
    if (widget.lang.length == 0) {
      return name;
    }
    if (widget.lang[name] != null) {
      return widget.lang[name];
    }
    return name;
  }

  void assetCallback(int index) {
    if (widget.onTap != null) {
      AssetEntity asset = media[index];
      widget.onTap!(asset);
    }
  }

  Future<void> imageList() async {
    media = await currentAlbums!.getAssetListPaged(currentPage, 60);
    if (media.isEmpty) return;

    for (var i = 0; i < media.length; i++) {
      // Uint8List? _thumb = await media[i].thumbData;
      Uint8List? _thumb = await media[i].thumbDataWithSize(200, 200);
      Widget asset;
      asset = InkWell(
        child: Container(
          margin: const EdgeInsets.all(0.5),
          child: Image.memory(
            _thumb!,
            fit: BoxFit.cover,
          ),
        ),
        onTap: () {
          assetCallback(i);
          Navigator.maybePop(context);
        },
      );
      _mediaList.add(asset);

      if (mounted) setState(() {});
    }
    currentPage++;
  }

  Widget _body() {
    if (_showAlbumsList) {
      if (albums != null) {
        return AlbumsList(
          albums!,
          onTap: (v) {
            currentPage = 0;
            lastPage = -1;
            _mediaList.clear();
            currentAlbums = v;
            imageList();
            if (mounted) setState(() => _showAlbumsList = false);
          },
        );
      }
      return Container();
    } else {
      return NotificationListener<ScrollNotification>(
        onNotification: (ScrollNotification scroll) {
          _loadMore(scroll);
          return false;
        },
        child: GridView.builder(
          itemCount: currentAlbums == null ? 0 : currentAlbums!.assetCount,
          gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
              crossAxisCount: widget.crossAxisCount),
          itemBuilder: (BuildContext context, int index) {
            if (_mediaList.length >= index + 1) {
              return _mediaList[index];
            }

            return Container(
              margin: const EdgeInsets.all(0.5),
              height: 30,
              color: Color(0xff4C4A48),
            );
          },
        ),
      );
    }
  }

  Widget get albumName {
    if (currentAlbums == null) {
      return Text('Loading...');
    }
    return Text(
      '${currentAlbums!.name}',
      style: TextStyle(
        height: 1.1,
        textBaseline: TextBaseline.alphabetic,
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black,
      appBar: AppBar(
        backgroundColor: Colors.black,
        leading: CloseButton(),
        leadingWidth: 30.0,
        title: InkWell(
          child: Container(
            height: 35.0,
            padding: EdgeInsets.only(left: 10.0, right: 10.0),
            decoration: BoxDecoration(
              color: Color(0xFF4C4C4C),
              borderRadius: new BorderRadius.circular((20.0)),
            ),
            child: Row(
              mainAxisAlignment: MainAxisAlignment.spaceAround,
              mainAxisSize: MainAxisSize.min,
              children: [
                albumName,
                Icon(
                  _showAlbumsList
                      ? Icons.keyboard_arrow_up
                      : Icons.keyboard_arrow_down,
                  size: 30.0,
                ),
              ],
            ),
          ),
          onTap: () {
            if (mounted) {
              setState(() => _showAlbumsList = !_showAlbumsList);
            }
          },
        ),
      ),
      body: _body(),
    );
  }
}
